### PR TITLE
Add summary data type to metricdata

### PR DESCRIPTION
**Pull request by dashpole**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/4587

This only adds the summary type to metricdata in the SDK, and does not add a corresponding AggregationSummary view aggregation.  This means that the only way to create these types is to use a bridge, and they cannot otherwise be created with the OpenTelemetry SDK.

### CODE REVIEW

File: sdk/metric/metricdata/data.go [hunk 1]
```diff
@@ -240,3 +240,54 @@ type Exemplar[N int64 | float64] struct {
...
+	// Count is the number of updates this summary has been calculated with.
+	Count uint64
+
+	// Sum is the sum of the values recorded.
+	Sum *float64
+
+	// (Optional) list of values at different quantiles of the distribution calculated
+	// from the current snapshot. The quantiles must be strictly increasing.
+	QuantileValues []ValueAtQuantile
+}
```
Comment (MrAlias) [inline sdk/metric/metricdata/data.go:280]: Does this not support exemplars?

File: sdk/metric/metricdata/data.go [hunk 2]
```diff
@@ -240,3 +240,54 @@ type Exemplar[N int64 | float64] struct {
...
+	// StartTime is when the timeseries was started.
+	StartTime time.Time
+	// Time is the time when the timeseries was recorded.
+	Time time.Time
+
+	// Count is the number of updates this summary has been calculated with.
+	Count uint64
+
+	// Sum is the sum of the values recorded.
+	Sum *float64
```
Comment (MrAlias) [inline sdk/metric/metricdata/data.go:275]: Can this ever be an `int64`? Should this be a generic type and this a generic value instead?

File: sdk/metric/metricdata/data.go [hunk 3]
```diff
@@ -240,3 +240,54 @@ type Exemplar[N int64 | float64] struct {
...
+	// from the current snapshot. The quantiles must be strictly increasing.
+	QuantileValues []ValueAtQuantile
+}
+
+// ValueAtQuantile the value at a given quantile of a distribution.
+type ValueAtQuantile struct {
+	// The quantile of a distribution.
+	//
+	// Must be in the interval [0.0, 1.0].
+	Quantile float64
```
Comment (MrAlias) [inline sdk/metric/metricdata/data.go:287]: nit: what do you think about calling this `QuantileValue`?

```suggestion
// QuantileValue the value at a given quantile of a distribution.
type QuantileValue struct {
	// Quantile is the quantile of this value.
	//
	// Must be in the interval [0.0, 1.0].
	Quantile float64
```

File: sdk/metric/metricdata/data.go (hunk 1 above)
Comment (dashpole) [reply sdk/metric/metricdata/data.go:280]: It does not: https://github.com/open-telemetry/opentelemetry-proto/blob/ac3242b03157295e4ee9e616af53b81517b06559/opentelemetry/proto/metrics/v1/metrics.proto#L584

File: sdk/metric/metricdata/data.go (hunk 2 above)
Comment (dashpole) [reply sdk/metric/metricdata/data.go:275]: It cannot.  I could "future-proof" this by making it generic with only float64 as an option.  WDYT?

Comment (dashpole) [reply sdk/metric/metricdata/data.go:275]: But the current proto for summaries only allows `double`: https://github.com/open-telemetry/opentelemetry-proto/blob/ac3242b03157295e4ee9e616af53b81517b06559/opentelemetry/proto/metrics/v1/metrics.proto#L617.  I do think it is incredibly unlikely to change.

[the remark this answers is not in the document]
Comment (dashpole) [reply sdk/metric/metricdata/data.go]: It is actually not optional in the proto, so i've removed the pointer.  It probably should have been optional (similar to histogram sum in the proto), but that ship has sailed.

File: sdk/metric/metricdata/data.go (hunk 3 above)
Comment (dashpole) [reply sdk/metric/metricdata/data.go]: updated

File: sdk/metric/metricdata/data.go (hunk 2 above)
Comment (MrAlias) [reply sdk/metric/metricdata/data.go]: Given the change required by the proto to support more than `double` would need to be backwards compatible as well, it seems like we can just follow the pattern here. SGTM :+1:

File: CHANGELOG.md [hunk 4]
```diff
@@ -16,6 +16,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add the `go.opentelemetry.io/otel/trace/noop` package as a default no-op implementation of the trace API. (#4620)
 - Add context propagation in `go.opentelemetry.io/otel/example/dice`. (#4644)
 - Add view configuration to `go.opentelemetry.io/otel/example/prometheus`. (#4649)
+- Add Summary, SummaryDataPoint, and ValueAtQuantile to `go.opentelemetry.io/sdk/metric/metricdata`. (#4622)
```
Comment (MrAlias) [inline CHANGELOG.md:19]:
```suggestion
- Add Summary, SummaryDataPoint, and QuantileValue to `go.opentelemetry.io/sdk/metric/metricdata`. (#4622)
```

**Review by MrAlias — APPROVED** [review]

**Pull request merged**